int16 x
int16 y
int16 d
int16 e